Add blake2/generichash oneshot implementation

lmctv opened the pull request:
formerly in #201

lmctv commented on src/nacl/bindings/crypto_generichash.py:
```diff
@@ -0,0 +1,50 @@
+                )
+    # _salt and _personal will be truncated or zero-padded
+    # to the correct length
+
```
Both parameters are defined as being of a fixed size and zero-filled.
This is the reason why I think the memmove usage to be correct.

lmctv commented:
@reaperhulk  I've changed my implementation so that it now mimics the behaviour of blake2b as implemented in python 3.6 by just zero padding, and raising an exception for too long salt and personal.

zookozcash commented:
I also read through these patches at the same time as https://github.com/pyca/pynacl/pull/217#issuecomment-266923156

lmctv commented:
@reaperhulk  I've splitted the PR into three parts, in the hope this will help reviews.

reaperhulk commented:
This can be rebased now

lmctv commented:
@reaperhulk rebased, as requested.

warner commented on src/nacl/bindings/crypto_generichash.py:
```diff
@@ -0,0 +1,72 @@
+def generichash_blake2b_salt_personal(data,
+                                      digest_size=crypto_generichash_BYTES,
+                                      key=b'', salt=b'', person=b''):
+    """One time hash interface"""
```
might be nice if the docstring said "one *shot* hash interface", instead of one-time

lmctv replied:
I think I wrote the docstring before thinking about the "one shot" phrase... Will change.

warner commented on src/nacl/hash.py:
```diff
@@ -17,10 +17,80 @@
     return encoder.encode(nacl.bindings.crypto_hash_sha512(message))
+
+
+def generichash(data, digest_size=BYTES, key=b'',
```
I'm dubious about whether such a generic-sounding name as `generichash` should be exposed. I know nacl likes to offer a short(er) name for the specific recommended algorithm for each class of operations, but I don't think hiding the specific hash being used serves anybody very well. Maybe we should leave this function out, and just offer `blake2b()` instead?

lmctv replied:
While for my own expected use cases I only need the salted and personalized version, I think feature-parity of the bindings with the C library can aid those coming to the python bindings from libsodium's documentation, which refers to the blake2 name (and API) only in a page about key derivation.

warner commented on src/nacl/hash.py:
```diff
@@ -17,10 +17,80 @@
+    :param salt: an initialization salt of lenght :py:data:`.SALTBYTES`.
+                 It will be zero-padded or truncated if needed.
+    :type salt: bytes
+    :param person: a personalization string of lenght
```
Any thoughts about calling this `personalization=` instead of `person=`? The latter sounds pretty weird.

lmctv replied:
I tried to implement the same call signature used in python3.6's stdlib version, modulo missing support for tree hashes in libsodium's version.

reaperhulk commented:
Looks like we need a new `hash.rst` in `docs` documenting this (see the public/secret/signing docs for a sample of what we want it to look like)

reaperhulk commented on src/nacl/bindings/crypto_generichash.py:
```diff
@@ -0,0 +1,72 @@
+    _salt = ffi.new("char []", crypto_generichash_SALTBYTES)
+    _person = ffi.new("char []", crypto_generichash_PERSONALBYTES)
+
+    ffi.memmove(_salt, salt, len(salt))
```
Let's not use `memmove` here. Users should pass the full length required or `None`. If they pass `None` we should pass `ffi.NULL` to the function rather than building the null byte string on our end (this goes for key, salt, and personalization).

lmctv replied:
For all the parameters key, salt, personal, I see the very same pattern in cpython implementation: respectively at lines 130, 141 and 228 of
https://github.com/python/cpython/blob/c30098c8c6014f3340a369a31df9c74bdbacc269/Modules/_blake2/blake2b_impl.c .

Here are a couple of test from cpython3.6's hashlib, where I've tried putting in short salt and personal, and the zero-padded version of the same:
```
>>> hashlib.blake2b(b'asdf', key=b'asd', salt=b'sdf', person=b'ply').hexdigest()
'931a7d96acc24a7954420d45404e5351465e00c45188341da66c0fe01254c8e5a65a320a2f399006afe56b60262fe918dff9ef5cd62eb7850b2f09d5bb12e364'
>>> hashlib.blake2b(b'asdf', key=b'asd', salt=b'sdf'+13*b'\x00', person=b'ply').hexdigest()
'931a7d96acc24a7954420d45404e5351465e00c45188341da66c0fe01254c8e5a65a320a2f399006afe56b60262fe918dff9ef5cd62eb7850b2f09d5bb12e364'
>>> hashlib.blake2b(b'asdf', key=b'asd', salt=b'sdf'+13*b'\x00', person=b'ply'+13*b'\x00').hexdigest()
'931a7d96acc24a7954420d45404e5351465e00c45188341da66c0fe01254c8e5a65a320a2f399006afe56b60262fe918dff9ef5cd62eb7850b2f09d5bb12e364'
```

lmctv replied:
I've just seen it should be possible avoiding the ``memmove`` for the key, since it's block should be zero-padded on input.

lmctv replied:
I didn't remember, but I've had removed the memmove in the first iterations of this PR.
The other two are operationally correct, and needed to avoid crashes under pypy.

lmctv commented:
There has been a connection failure while executing Job #560.11  https://travis-ci.org/pyca/pynacl/jobs/189626734 .
It would be useful if someone with admin access to travis could restart just the failed job.

lmctv commented:
Travis' network hates me!  In build job #565.2 -  https://travis-ci.org/pyca/pynacl/jobs/189742687  - while

```
Downloading virtualenv-15.1.0-py2.py3-none-any.whl (1.8MB)
    45% |██████████████▋                 | 829kB 10.6MB/s eta 0:00:01Exception:
```

there has been a disconnection:
```
Traceback (most recent call last):
   [...]
  File "/home/travis/.pyenv/versions/3.3.6/lib/python3.3/ssl.py", line 352, in read
    v = self._sslobj.read(len, buffer)
ConnectionResetError: [Errno 104] Connection reset by peer
```

lmctv commented:
@warner I've added a proper docstring for the low-level python bindings, and corrected that "one time" in "one shot" as suggested; as for nacl.hash.generichash , now it's simply another name for nacl.hash.blake2b instead of a mostly copy-pasted implementation.

Could you do another review pass?

lmctv commented:
@warner , @reaperhulk  I've added a little bit of documentation, as requested by Paul, and I think I've already complied with Brian's requests last week. Would you mind one more review pass?

reaperhulk commented on tests/test_generichash.py:
```diff
@@ -0,0 +1,109 @@
+def _read_crypto_test_vectors(fname):
+    vectors = []
+    path = os.path.join(os.path.dirname(__file__), "data", fname)
+    with open(path, "r") as fp:
```
Open this with `rb` and then you shouldn't need the `bytes_as_string` helper since on py3 the resulting data will be a bytestring as expected.

lmctv replied:
I've just tried to implement your suggestion, but, while all goes well on python2.7, all I've got on python3.5 is a crash, deep inside py.test during test collection, before the real test could even start:

```
==================================== ERRORS ====================================
__________________ ERROR collecting tests/test_generichash.py __________________
venv-35/lib/python3.5/site-packages/_pytest/runner.py:163: in __init__
    self.result = func()
venv-35/lib/python3.5/site-packages/_pytest/main.py:464: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
venv-35/lib/python3.5/site-packages/_pytest/main.py:335: in _memoizedcall
    res = function()
# # # [  .  .  .  ]
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
venv-35/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:614: in execute
    res = hook_impl.function(*args)
venv-35/lib/python3.5/site-packages/_pytest/python.py:110: in pytest_generate_tests
    metafunc.parametrize(*marker.args, **marker.kwargs)
venv-35/lib/python3.5/site-packages/_pytest/python.py:856: in parametrize
    assert len(valset) == len(argnames)
E   AssertionError
```

I think it's better to leave things as they stand now.

lmctv replied:
OK, I've managed to tame python3.5 bytes comparison, and therefore added a commit removing `bytes_as_string` usage as you requested.

I'm unsure if doing a more drastic rebase to remove the helper's definition from history and having it moved to the init/update/finalize PR, or leaving the commit in for (near?) future use.

reaperhulk replied:
You can leave it here and we can discuss whether it needs to exist in the init/update/finalize PR :)

reaperhulk commented:
Since the next PR will immediately address the function that doesn't have coverage I'm merging.

lmctv commented:
@reaperhulk @warner  thank you very much for your help!
